design/service-apis-implementation.md

```diff
@@ -0,0 +1,342 @@
+So, there are some features that are currently gaps for Contour (wildcard domain names, and exact path matching seem like big ones),
+and some that Contour supports that the Service APIs do not (websockets, configurable timeouts, header replacement, external auth, rate limiting, and so on).
+
+Our eventual ideal is that HTTPProxy and Service-APIs will have feature parity.
+Practically, there may be times when HTTPProxy can move faster than Service-APIs because we have a smaller feature surface.
+We see this as a chance for HTTPProxy to test out ideas for Service-APIs functionality, and provide feedback on what features should be in the core, extension, and implementation-specific parts of the API.
+There should never be annotations required for functionality on Service-APIs objects.
+
+
+## Goals
+- Define a data model and implementation for Contour's service-apis support, covering the v1alpha1 version of the service-apis.
+- Layer 7 support only, which means HTTPRoutes, TLSRoutes only.
+
+## Non Goals
+- No TCPRoute or UDPRoute support, that is, no support for arbitrary TCP/UDP forwarding.
+
+
+## High-Level Design
+
+### Service APIs implementation surface area
+
+Contour's support for the Service APIs will include support for the HTTPRoute and TLSRoute objects only, not any other type of Route - this includes no TCPRoute or UDPRoute support.
+Contour is a layer 7 ingress controller, and the layer 4 load balancing implies by TCPRoute and UDPRoute is out of scope for this tool (that is, `projectcontour/contour`).
+Project Contour (the organisation) may investigate a Service-APIs based layer 4 solution in the future, but that effort will not be in `projectcontour/contour`.
+
+### Contour Gateway model
+
+Contour considers a Gateway to describe a single Envoy deployment, deployed and managed by something outside of itself.
+Contour expects to be supplied with the full name (name and namespace) of a single Gateway, which it will watch and update the status of.
+This will be a configuration file parameter.
+When a Gateway is not supplied, Contour will not do Service-APIs processing, or in other words, the existing `--experimental-service-apis` flag will be obsoleted and removed).
+Inside Contour will merge Listeners within its Gateway. (See the detailed design for the exact rules Contour will use for this.)
+
+### Interrelated watches
+
+The Service APIs are a set of interrelated Kubernetes objects, where a change in one object can mean that the scope of objects Contour is interested in will change.
+Because of this, Contour will watch all events associated with its named Gateway, HTTPRoute, TLSRoute, and BackendPolicy objects, and filter the objects it takes action on internally, rather than using a filtered watch (as those are costly to set up and tear down).
+
+Contour does not watch a GatewayClass, as its deployment model expects a Contour to be part of the implementation for a *particular* Gateway.
```

Comment:
If we did watch the class, we could allow a single gateway in a cluster to be configured by a single contour without any configuration file changes. But when a second Gateway for the same class controller was created you'd need to specify.

Would just be nice to allow users to configure Gateways without having to edit the configmap + restart contour instances.

Reply:
I think that it's better to be explicit for now. We can revisit this later if the configuration proves too complicated. This way we can avoid a more complicated state like we had for the ingress.class annotation, where the behavior if you don't specify a value is not straightforward.